From 85917943719772b25a6f0fd5465427665889c52f Mon Sep 17 00:00:00 2001
From: Ken Ma <make@marvell.com>
Date: Mon, 15 May 2017 14:14:13 +0800
Subject: [PATCH 1009/1345] fix: comphy: a3700: align comphy lane number with
 func spec

commit  5fe80b236700ccc341a481522930f8c6737adabc from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

In current linux dts and driver, lane0 is for PCIe/SGMII0 and lane1 is
for USB3/SGMII1 which opposites the A3700 func spec;
This patch aligns lane number with the spec - lane0 is for USB3/SGMII1
and lane1 is for PCIe/SGMII0.

Change-Id: If5d8fe0f6728f8a65d16bb900aff0b08bf03a74c
Signed-off-by: Ken Ma <make@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/39549
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Victor Gu <xigu@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/39673
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../boot/dts/marvell/armada-3720-community.dts     |    6 +++---
 arch/arm64/boot/dts/marvell/armada-3720-db-B.dts   |    2 +-
 arch/arm64/boot/dts/marvell/armada-3720-db.dts     |    6 +++---
 drivers/phy/phy-comphy-a3700.c                     |   22 ++++++++++----------
 drivers/phy/phy-comphy-a3700.h                     |    4 ++--
 5 files changed, 20 insertions(+), 20 deletions(-)

diff --git a/arch/arm64/boot/dts/marvell/armada-3720-community.dts b/arch/arm64/boot/dts/marvell/armada-3720-community.dts
index 2c60fa1..8035255 100644
--- a/arch/arm64/boot/dts/marvell/armada-3720-community.dts
+++ b/arch/arm64/boot/dts/marvell/armada-3720-community.dts
@@ -112,7 +112,7 @@
 
 			u3d@50000 {
 				status = "disabled";
-				phys = <&a3700_comphy 1 COMPHY_USB3>;
+				phys = <&a3700_comphy 0 COMPHY_USB3>;
 				phy-names = "usb";
 			};
 
@@ -126,7 +126,7 @@
 				compatible = "generic-xhci";
 				status = "okay";
 				phys = <&utmi_usb32>, /* utmi usb32 dedicated phy */
-				       <&a3700_comphy 1 COMPHY_USB3>; /* usb3 comphy */
+				       <&a3700_comphy 0 COMPHY_USB3>; /* usb3 comphy */
 				phy-names = "usb2", "usb3";
 			};
 
@@ -209,7 +209,7 @@
 
 &pcie0 {
 	status = "okay";
-	phys = <&a3700_comphy 0 COMPHY_PCIE0>;
+	phys = <&a3700_comphy 1 COMPHY_PCIE0>;
 	phy-names = "comphy";
 	reset-gpios = <&gpio_sb 3 GPIO_ACTIVE_HIGH>;
 };
diff --git a/arch/arm64/boot/dts/marvell/armada-3720-db-B.dts b/arch/arm64/boot/dts/marvell/armada-3720-db-B.dts
index c8db25c..c58d641 100644
--- a/arch/arm64/boot/dts/marvell/armada-3720-db-B.dts
+++ b/arch/arm64/boot/dts/marvell/armada-3720-db-B.dts
@@ -54,7 +54,7 @@
 			eth1: ethernet@40000 {
 				phy-mode = "sgmii";
 				phy = <&phy1>;
-				phys = <&a3700_comphy 1 COMPHY_SGMII1>;
+				phys = <&a3700_comphy 0 COMPHY_SGMII1>;
 				phy-names = "comphy";
 				status = "okay";
 				buffer-manager = <&bm>;
diff --git a/arch/arm64/boot/dts/marvell/armada-3720-db.dts b/arch/arm64/boot/dts/marvell/armada-3720-db.dts
index 49012a7..43e0ef0 100644
--- a/arch/arm64/boot/dts/marvell/armada-3720-db.dts
+++ b/arch/arm64/boot/dts/marvell/armada-3720-db.dts
@@ -90,7 +90,7 @@
 
 			u3d@50000 {
 				status = "okay";
-				phys = <&a3700_comphy 1 COMPHY_USB3>;
+				phys = <&a3700_comphy 0 COMPHY_USB3>;
 				phy-names = "usb";
 			};
 
@@ -274,7 +274,7 @@
 	status = "okay";
 	usb-phy = <&usb_phy>;
 	phys = <&utmi_usb32>, /* utmi usb32 dedicated phy */
-       <&a3700_comphy 1 COMPHY_USB3>; /* usb3 comphy */
+	<&a3700_comphy 0 COMPHY_USB3>; /* usb3 comphy */
 	phy-names = "usb2", "usb3";
 };
 
@@ -285,7 +285,7 @@
 	pinctrl-0 = <&pcie_pins>;
 
 	status = "okay";
-	phys = <&a3700_comphy 0 COMPHY_PCIE0>;
+	phys = <&a3700_comphy 1 COMPHY_PCIE0>;
 	phy-names = "comphy";
 };
 
diff --git a/drivers/phy/phy-comphy-a3700.c b/drivers/phy/phy-comphy-a3700.c
index d7a202d..84f7b42 100644
--- a/drivers/phy/phy-comphy-a3700.c
+++ b/drivers/phy/phy-comphy-a3700.c
@@ -42,9 +42,9 @@ static void mvebu_a3700_comphy_set_phy_selector(struct mvebu_comphy_priv *priv,
 
 	case (COMPHY_SGMII_MODE):
 	case (COMPHY_HS_SGMII_MODE):
-		if (comphy->index == COMPHY_LANE1)
+		if (comphy->index == COMPHY_LANE0)
 			reg &= ~COMPHY_SELECTOR_USB3_GBE1_SEL_BIT;
-		else if (comphy->index == COMPHY_LANE0)
+		else if (comphy->index == COMPHY_LANE1)
 			reg &= ~COMPHY_SELECTOR_PCIE_GBE0_SEL_BIT;
 		else
 			dev_err(priv->dev, "COMPHY[%d] mode[%d] is invalid\n", comphy->index, mode);
@@ -55,15 +55,15 @@ static void mvebu_a3700_comphy_set_phy_selector(struct mvebu_comphy_priv *priv,
 	case (COMPHY_USB3_MODE):
 		if (comphy->index == COMPHY_LANE2)
 			reg |= COMPHY_SELECTOR_USB3_PHY_SEL_BIT;
-		else if (comphy->index == COMPHY_LANE1)
+		else if (comphy->index == COMPHY_LANE0)
 			reg |= COMPHY_SELECTOR_USB3_GBE1_SEL_BIT;
 		else
 			dev_err(priv->dev, "COMPHY[%d] mode[%d] is invalid\n", comphy->index, mode);
 		break;
 
 	case (COMPHY_PCIE_MODE):
-		/* PCIE must be in Lane0 */
-		if (comphy->index == COMPHY_LANE0)
+		/* PCIE must be in Lane1 */
+		if (comphy->index == COMPHY_LANE1)
 			reg |= COMPHY_SELECTOR_PCIE_GBE0_SEL_BIT;
 		else
 			dev_err(priv->dev, "COMPHY[%d] mode[%d] is invalid\n", comphy->index, mode);
@@ -285,10 +285,10 @@ static int mvebu_a3700_comphy_sgmii_power_on(struct mvebu_comphy_priv *priv,
 	mvebu_a3700_comphy_set_phy_selector(priv, comphy);
 
 	/* Serdes IP Base address
-	 * COMPHY Lane0 -- PCIe/GBE0
-	 * COMPHY Lane1 -- USB3/GBE1
+	 * COMPHY Lane0 -- USB3/GBE1
+	 * COMPHY Lane1 -- PCIe/GBE0
 	 */
-	if (comphy->index == COMPHY_LANE1) {
+	if (comphy->index == COMPHY_LANE0) {
 		/* Get usb3 and gbe register resource and map */
 		res = platform_get_resource_byname(pdev, IORESOURCE_MEM, "usb3_gbe1_phy");
 		if (res) {
@@ -461,7 +461,7 @@ static int mvebu_a3700_comphy_sgmii_power_on(struct mvebu_comphy_priv *priv,
 		dev_err(priv->dev, "Failed to init RX of SGMII PHY %d\n", comphy->index);
 
 	/* Unmap resource */
-	if (comphy->index == COMPHY_LANE1) {
+	if (comphy->index == COMPHY_LANE0) {
 		devm_iounmap(&pdev->dev, sd_ip_addr);
 		devm_release_mem_region(&pdev->dev, res->start, resource_size(res));
 	}
@@ -1003,9 +1003,9 @@ static int mvebu_a3700_comphy_is_pll_locked(struct phy *phy)
 	.num_of_lanes = 3,
 	.functions = {
 		/* Lane 0 */
-		{COMPHY_UNUSED, COMPHY_PCIE0, COMPHY_SGMII0},
-		/* Lane 1 */
 		{COMPHY_UNUSED, COMPHY_SGMII1, COMPHY_HS_SGMII1, COMPHY_USB3},
+		/* Lane 1 */
+		{COMPHY_UNUSED, COMPHY_PCIE0, COMPHY_SGMII0},
 		/* Lane 2 */
 		{COMPHY_UNUSED, COMPHY_SATA0, COMPHY_USB3},
 	},
diff --git a/drivers/phy/phy-comphy-a3700.h b/drivers/phy/phy-comphy-a3700.h
index 7886aa9..becf697 100644
--- a/drivers/phy/phy-comphy-a3700.h
+++ b/drivers/phy/phy-comphy-a3700.h
@@ -205,7 +205,7 @@ enum {
 #define CFG_PM_RXDLOZ_WAIT_12_UNIT		(0xC << CFG_PM_RXDLOZ_WAIT_OFF)
 
 /* SGMII */
-#define COMPHY_PHY_CFG1_OFFSET(lane)		((lane) * 0x28)
+#define COMPHY_PHY_CFG1_OFFSET(lane)		((1 - (lane)) * 0x28)
 #define PIN_PU_IVEREF_BIT			BIT(1)
 #define PIN_RESET_CORE_BIT			BIT(11)
 #define PIN_RESET_COMPHY_BIT			BIT(12)
@@ -225,7 +225,7 @@ enum {
  * lane0: PCIe/GbE0 PHY Status 1
  * lane1: USB3/GbE1 PHY Status 1
  */
-#define COMPHY_PHY_STATUS_OFFSET(lane)		(0x18 + (lane) * 0x28)
+#define COMPHY_PHY_STATUS_OFFSET(lane)		(0x18 + (1 - (lane)) * 0x28)
 #define PHY_RX_INIT_DONE_BIT			BIT(0)
 #define PHY_PLL_READY_RX_BIT			BIT(2)
 #define PHY_PLL_READY_TX_BIT			BIT(3)
-- 
1.7.9.5

